numba/targets/removerefctpass.py

```diff
@@ -0,0 +1,109 @@
+"""
+Implement a rewrite pass on LLVM module to remove unnecessary refcount
+operation.
+"""
+from __future__ import absolute_import, print_function
+
+from llvmlite.ir.transforms import CallVisitor
+
+from numba import types
+
+
+class _MarkNrtCallVisitor(CallVisitor):
+    """
+    A pass to mark all NRT_incref and NRT_decref.
+    """
+    def __init__(self):
+        self.marked = set()
+
+    def visit_Call(self, instr):
+        if instr.callee.name in ('NRT_incref', 'NRT_decref'):
+            self.marked.add(instr)
+
+
+def _rewrite_function(function):
+    # Mark NRT usage
+    markpass = _MarkNrtCallVisitor()
+    markpass.visit_Function(function)
+    marked = markpass.marked
+    # Remove NRT usage
+    for bb in function.basic_blocks:
+        for inst in list(bb.instructions):
+            if inst in marked:
+                bb.instructions.remove(inst)
+
+
+_accepted_nrtfns = 'NRT_incref', 'NRT_decref'
+
+
+def _legalize(module, dmm, fndesc):
+    """
+    Legalize the code in the module.
+    Returns True if the module is legal for the rewrite pass that remove
+    unnecessary refcount.
+    """
+
+    def valid_output(ty):
+        """
+        Valid output are any type that does not need refcount
+        """
+        model = dmm[ty]
+        return not model.contains_nrt_meminfo()
```

Comment:
Would a function that returns one of its array arguments also be a valid candidate for this rewrite pass?